docs/react/guides/advanced-ssr.md

```diff
@@ -0,0 +1,389 @@
+import {
+  dehydrate,
+  HydrationBoundary,
+  QueryClient,
+} from '@tanstack/react-query'
+import Posts from './posts'
+
+export default async function PostsPage() {
+  const queryClient = new QueryClient()
+
+  // Note we are now using fetchQuery()
+  const posts = await queryClient.fetchQuery({
+    queryKey: ['posts'],
+    queryFn: getPosts,
+  })
+
+  return (
+    <HydrationBoundary state={dehydrate(queryClient)}>
+      {/* This is the new part */}
+      <div>Nr of posts: {posts.length}</div>
+      <Posts />
+    </HydrationBoundary>
+  )
+}
+```
+
+We are now rendering data from the `getPosts` query both in a Server Component and in a Client Component. This will be fine for the initial page render, but what happens when the query revalidates on the client for some reason when `staleTime` has been passed?
+
+React Query has no idea of how to _revalidate the Server Component_, so if it refetches the data on the client, causing React to rerender the list of posts, the `Nr of posts: {posts.length}` will end up out of sync.
+
+This is fine if you set `staleTime: Infinity`, so that React Query never revalidates, but this is probably not what you want if you are using React Query in the first place.
+
+Using React Query with Server Components makes most sense if:
+
+- You have an app using React Query and want to migrate to Server Components without rewriting all the data fetching
+- You want a familiar programming paradigm, but want to still sprinkle in the benefits of Server Components where it makes most sense
+- You have some use case that React Query covers, but that your framework of choice does not cover
+
+It's hard to give general advice on when it makes sense to pair React Query with Server Components and not. **If you are just starting out with a new Server Components app, we suggest you start out with any tools for data fetching your framework provides you with and avoid bringing in React Query until you actually need it.** This might be never, and that's fine, use the right tool for the job!
+
+If you do use it, a good rule of thumb is to avoid `queryClient.fetchQuery` unless you need to catch errors. If you do use it, don't render its result on the server or pass the result to another component, even a Client Component one.
+
+From the React Query perspective, treat Server Components as a place to prefetch data, nothing more.
+
+Of course, it's fine to have Server Components own some data, and Client Components own other, just make sure those two realities don't get out of sync.
+
+## Streaming with Server Components
+
+The Next.js app router automatically streams any part of the application that is ready to be displayed to the browser as soon as possible, so finished content can be displayed immediately without waiting for still pending content. It does this along `<Suspense>` boundary lines. Note that if you create a file `loading.tsx`, this automatically creates a `<Suspense>` boundary behind the scenes.
```

Comment:
I've seen that we get an infinite fetch loop on the client if we don't have a suspense boundary with our experimental streaming plugin.

Not sure if it works with a `loading.tsx`, but maybe we should just recommend to always set a suspense boundary? you'd need it for client transitions anyways I guess!

Reply:
Good idea mentioning that at least for the streaming plugin, I added a note there. I think it's harder for us to have recommendations for Suspense boundaries in general though because they are so framework and application specific. It really matters where they go etc. I don't you necessarily need need one for the general approach described on this page, the reason the streaming lib is different is because it happens during the SSR phase, not the RSC phase.